tensorboard/uploader/proto/blob.proto

```diff
@@ -12,3 +12,14 @@ enum BlobState {
   // Object is finalized.
   CURRENT = 2;
 }
+
+message Blob {
+  // For Blobs that haven't started uploading, this is an empty string.
+  string blob_id = 1;
```

Comment:
(Copied from internal discussion.) This doesn’t feel right to me. `Blob`
and `BlobSequence` are not just portions of responses from particular
export RPCs. They’re resources: <https://aip.dev/121>. Resources are
supposed to be well defined by themselves, and have a hierarchical
relationship. As written here, a `BlobSequence` could contain two copies
of `Blob(id="", state=UNFINALIZED)` to represent two different holes.
But those two `Blob` entries don’t really represent blobs—just the
opposite!—and they’re certainly not “the same thing”, despite being
structurally identical.

If we were designing this in a sane type system, surely we would write
something like:

```rust
struct BlobSequence {
  blobs: Vec<Option<Blob>>,
}
struct Blob {
  id: Uuid,  // not optional, not empty
  state: BlobState,
}
enum BlobState { Unfinalized, Current }
```

The simplest way to express that in the proto IDL is something like:

```proto
message BlobSequence {
  repeated BlobSequenceEntry entries = 1;
}
message BlobSequenceEntry {
  // Optional. If absent, this represents a "hole" in the sequence:
  // there is expected to be a blob here, but upload has not started.
  Blob blob = 1;
}
message Blob {
  string id = 1;
  BlobState state = 2;
}
```

This way, any `Blob` actually represents a blob, rather than just
coöpting the bit patterns of the `Blob` struct to encode something
totally different. Can we adopt this structure instead?

Reply:
OK. The suggested way of dealing with holes n the blob sequence makes sense to me. Adopted with minor modifications (doc strings and ordering).

Reply:
Of course—thanks!